File: azure_function/Models/AnalysisResult.cs
using System.Text.Json.Serialization;

namespace EcoMonitorFunction.Models;

/// <summary>
/// Resultado completo del análisis de huella de carbono devuelto por CarbonFootprintAnalyzer.
/// </summary>
public class AnalysisResult
{
    [JsonPropertyName("periodo")]
    public PeriodoInfo Periodo { get; set; } = new();

    [JsonPropertyName("resumen")]
    public ResumenLecturas Resumen { get; set; } = new();

    [JsonPropertyName("analisis_ia")]
    public string AnalisisIa { get; set; } = string.Empty;

    [JsonPropertyName("veredicto")]
    public string Veredicto { get; set; } = string.Empty;

    [JsonPropertyName("timestamp_analisis")]
    public DateTime TimestampAnalisis { get; set; } = DateTime.UtcNow;
}

public class PeriodoInfo
{
    [JsonPropertyName("desde")]
    public string Desde { get; set; } = string.Empty;

    [JsonPropertyName("hasta")]
    public string Hasta { get; set; } = string.Empty;

    [JsonPropertyName("horas")]
    public double Horas { get; set; }
}

public class ResumenLecturas
{
    [JsonPropertyName("total_lecturas")]
    public int TotalLecturas { get; set; }

    [JsonPropertyName("total_kwh")]
    public double TotalKwh { get; set; }

    [JsonPropertyName("total_co2_kg")]
    public double TotalCo2Kg { get; set; }

    [JsonPropertyName("co2_por_dispositivo")]
    public Dictionary<string, double> Co2PorDispositivo { get; set; } = new();

    [JsonPropertyName("kwh_por_dispositivo")]
    public Dictionary<string, double> KwhPorDispositivo { get; set; } = new();

    [JsonPropertyName("factor_emisión_usado")]
    public double FactorEmisionUsado { get; set; }

    [JsonPropertyName("num_dispositivos")]
    public int NumDispositivos { get; set; }
}

File: dashboard_angular/src/app/core/models/analysis-result.ts
/**
 * Modelos que reflejan la respuesta de la Azure Function CarbonFootprintAnalyzer
 * (GET /api/analyze). Los nombres de campo coinciden con el JSON del backend.
 */
export interface Periodo {
  desde: string;
  hasta: string;
  horas: number;
}

export interface Resumen {
  total_lecturas: number;
  total_kwh: number;
  total_co2_kg: number;
  co2_por_dispositivo: Record<string, number>;
  kwh_por_dispositivo: Record<string, number>;
  'factor_emisión_usado': number;
  num_dispositivos: number;
}

export interface AnalysisResult {
  periodo: Periodo;
  resumen: Resumen;
  analisis_ia: string;
  veredicto: string;
  timestamp_analisis: string;
}

/** Respuesta cuando no hay lecturas en el período. */
export interface NoDataResponse {
  mensaje: string;
  desde: string;
  hasta: string;
  source?: string;
}

export type AnalyzeResponse = AnalysisResult | NoDataResponse;

/** Type guard: distingue una respuesta con datos de una respuesta vacía. */
export function tieneDatos(r: AnalyzeResponse): r is AnalysisResult {
  return (r as AnalysisResult).analisis_ia !== undefined;
}
